src/Compiler/Driver/OptimizeInputs.fs

```diff
@@ -48,6 +51,288 @@ let GetInitialOptimizationEnv (tcImports: TcImports, tcGlobals: TcGlobals) =
+                Phase3 = None
+            }
+
+    type Node =
+        {
+            Idx: int
+            Phase: OptimizationPhase
+        }
+
+        override this.ToString() = $"[{this.Idx}-{this.Phase}]"
+
+    let collectResults
+        (inputs: FileResultsComplete[])
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let files =
+            inputs
+            |> Array.map
+                (fun {
+                         Phase1 = phase1
+                         Phase2 = _phase2
+                         Phase3 = phase3
+                     } ->
+                    let (_, _, implFileOptData, _), optimizeDuringCodeGen = phase1
+                    let _, implFile = phase3
+
+                    let implFile =
+                        {
+                            ImplFile = implFile
+                            OptimizeDuringCodeGen = optimizeDuringCodeGen
+                        }
+
+                    implFile, implFileOptData)
+
+        let lastFilePhase1Env =
+            inputs
+            |> Array.last
+            |> fun { Phase1 = phase1 } ->
+                let (optEnvPhase1, _, _, _), _ = phase1
+                optEnvPhase1
+
+        files, lastFilePhase1Env
+
+    let getPhase2Res (p: FileResults) = p.Phase2 |> Option.get
+
+    let getPhase3Res (p: FileResults) =
+        p.Phase3 |> Option.get |> (fun (env, _) -> env)
+
+    let optimizeFilesInParallel
+        (env0: IncrementalOptimizationEnv)
+        ((phase1, phase2, phase3): FilePhaseFuncs)
+        (files: CheckedImplFile list)
+        (ct: CancellationToken)
+        : (CheckedImplFileAfterOptimization * ImplFileOptimizationInfo)[] * IncrementalOptimizationEnv =
+        let files = files |> List.toArray
+
+        let firstNode =
+            {
+                Idx = 0
+                Phase = OptimizationPhase.Phase1
+            }
+
+        let results = files |> Array.map (fun _ -> FileResults.Empty)
+
+        let _lock = obj ()
+
+        let nodeCanBeProcessed { Idx = idx; Phase = phase } : bool =
+            let previousPhase =
+                match phase with
+                | OptimizationPhase.Phase1 -> None
+                | OptimizationPhase.Phase2 -> Some OptimizationPhase.Phase1
+                | OptimizationPhase.Phase3 -> Some OptimizationPhase.Phase2
+
+            lock _lock (fun () ->
+                let previousFileReady = if idx = 0 then true else results[ idx - 1 ].HasResult phase
+
+                let previousPhaseReady =
+                    match previousPhase with
+                    | Some previousPhase -> results[ idx ].HasResult previousPhase
+                    | None -> true
+
+                previousFileReady && previousPhaseReady)
+
+        let visited = HashSet<Node>()
+
+        let worker ({ Idx = idx; Phase = phase } as node: Node) : Node[] =
+            let notPreviouslyVisited = lock _lock (fun () -> visited.Add node)
```

Comment:
The operation on `visited `is only locked-on here, but not on reading. That might bring inconsistent reasults if readers (which are not locked) attempt to read while the collection is being rebuilt.

BUT: In thise case it seems to work "just" as a counter, if I am not missing anything?
So either Interlocker.Increment (if it is just a counter) or a ConcurrentDictionairy would do it, therefore avoiding the potentially risky explicit locking approach. Risky because a user of the hashset might not realize locking is needed.

Reply:
I think I'll replace the more generic "graph"-based processing with 3 explicit worker tasks that run optimization for each phase, file-by-file. That will be much simpler.

Reply:
Agree, that would be best in making the intent clear.

Reply:
Have replaced with a 2D array of Tasks and task CEs with no locks.